runtime/domain.c

```diff
@@ -1646,27 +1645,7 @@ Caml_inline void advance_global_major_slice_epoch (caml_domain_state* d)
   if (old_value != atomic_load (&caml_minor_collections_count)) {
     /* This domain is the first one to use up half of its minor heap arena
         in this minor cycle. Trigger major slice on other domains. */
-    if (caml_plat_try_lock(&all_domains_lock)) {
-      /* Note that this interrupt is best-effort. If we get the lock,
-         then interrupt all the domains. If not, either some other domain
-         is calling for a stop-the-world section interrupting all the
-         domains, or a domain is being created or terminated. All of these
-         actions also try to lock [all_domains_lock] mutex, and the above
-         lock acquisition may fail.
-
-         If we don't get the lock, we don't interrupt other domains. This is
-         acceptable since it does not affect safety but only liveness -- the
-         speed of the major gc. The other domains may themselves fill half of
-         their minor heap triggering a major slice, or will certainly do a
-         major slice right after their next minor GC when they observe that
-         their domain-local [Caml_state->major_slice_epoch] is less than the
-         global one [caml_major_slice_epoch]. */
-      for(int i = 0; i < stw_domains.participating_domains; i++) {
-        dom_internal * di = stw_domains.domains[i];
-        if (di->state != d) interrupt_domain(&di->interruptor);
-      }
-      caml_plat_unlock (&all_domains_lock);
-    }
+    caml_interrupt_all_signal_safe();
```

Comment:
I want to add a note that this changes which domains are examined for interruption. While the current logic uses `stw_domains.participating_domains`, `caml_interrupt_all_signal_safe` examines the `all_domains` array whose size is `Max_domains`. This change will examine more domains than the current solution. I don't anticipate this being expensive in practice as this is done once per minor cycle. As such benchmarking isn't necessary. 

I have a question about `caml_interrupt_all_signal_safe`? It interrupts the domain by setting its `young_limit` to a large value `UINTNAT_MAX`. No lock is taken when doing this. Does that imply that it is always safe to set a domain's `young_limit` to `UINTNAT_MAX` irrespective of what it is doing?

Reply:
To clarify, the loop does not examine `Max_domains` domains, thanks to the early return. It examines n+1 domains where n is the maximum number for domains running at the same time in the current run of the program.

Why the invariant enabling this early return holds might not be obvious from the code, so I have made it more manifest with the latest commit.

Reply:
> Does that imply that it is always safe to set a domain's young_limit to UINTNAT_MAX irrespective of what it is doing?

Yes, in particular the domain should only set its `young_limit` by calling `caml_reset_young_limit`.